From f3dcf6b0eb140bd52d72e4c8d1998670a723c316 Mon Sep 17 00:00:00 2001
From: Zac Crosby <zac@motorola.com>
Date: Mon, 13 Jul 2015 09:21:11 -0500
Subject: [PATCH 174/959] IKKRNBSP-3415 msm: mdss: Print out DSI read registers
 in error cases

Print out the DSI read registers in the following cases:
- Invalid DSI command when receiving from the target
- ACK/ERR and data packet

Port of 3e487fd on 8996.

Change-Id: Ifc1ba7d17ac5ce385ef834e5899369aa389ba232
Signed-off-by: Shiyong Li <a22381@motorola.com>
Reviewed-on: http://gerrit.mot.com/786055
SLTApproved: Slta Waiver <sltawvr@motorola.com>
SME-Granted: SME Approvals Granted
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Zachary Crosby <zac@motorola.com>
Reviewed-by: Joseph Swantek <jswantek@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/video/msm/mdss/mdss_dsi_host.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/drivers/video/msm/mdss/mdss_dsi_host.c b/drivers/video/msm/mdss/mdss_dsi_host.c
index 3e9d2b5879b..d8b1d7e2ce8 100644
--- a/drivers/video/msm/mdss/mdss_dsi_host.c
+++ b/drivers/video/msm/mdss/mdss_dsi_host.c
@@ -1669,6 +1669,17 @@ do_send:
 	return len;
 }
 
+static void mdss_dsi_cmd_rx_data_log(struct mdss_dsi_ctrl_pdata *ctrl)
+{
+	pr_warn("%s: read_cnt = %d, DATA3 = 0x%08x, DATA2 = 0x%08x, DATA1 = 0x%08x, DATA0 = 0x%08x\n",
+		__func__,
+		MIPI_INP((ctrl->ctrl_base) + 0x01d4) >> 16,
+		MIPI_INP((ctrl->ctrl_base) + 0x078),
+		MIPI_INP((ctrl->ctrl_base) + 0x074),
+		MIPI_INP((ctrl->ctrl_base) + 0x070),
+		MIPI_INP((ctrl->ctrl_base) + 0x06c));
+}
+
 /* MIPI_DSI_MRPS, Maximum Return Packet Size */
 static char max_pktsize[2] = {0x00, 0x00}; /* LSB tx first, 10 bytes */
 
@@ -1900,7 +1911,8 @@ skip_max_pkt_size:
 		mdss_dsi_long_read_resp(rp);
 		break;
 	default:
-		pr_warning("%s:Invalid response cmd\n", __func__);
+		pr_warn("%s:Invalid response cmd=0x%x\n", __func__, cmd);
+		mdss_dsi_cmd_rx_data_log(ctrl);
 		rp->len = 0;
 		rp->read_cnt = 0;
 	}
@@ -2095,6 +2107,14 @@ static int mdss_dsi_cmd_dma_rx(struct mdss_dsi_ctrl_pdata *ctrl,
 		rp->read_cnt = (max_pktsize[0] + 6);
 	}
 
+	/* Sometimes target/display responds with the MIPI DSI acknowledge
+	 * and Error report after returning MIPI DSI read value, that cause
+	 * mdss_dsi_cmd_dma_rx() fails to parse. Print out all read registers
+	 * to investigate this issue
+	 */
+	if (ack_error && (rx_byte == 4))
+		mdss_dsi_cmd_rx_data_log(ctrl);
+
 	/*
 	 * In case of multiple reads from the panel, after the first read, there
 	 * is possibility that there are some bytes in the payload repeating in
-- 
2.11.0

